Introduce soc validator

nolash opened the pull request:
Introduces validation for single-owner chunks as per specifications in the book of swarm.

A new package `soc` is introduced which encapsulates a soc `Update` that works with or without signer, and theoretically can be used as a library to build and sign soc chunks.

It also dissolves the `validation` package, moving the `soc` validator to the `soc` package, and introducing the `content` package for both validator and convenience functions for _content addressed chunks_.

pradovic commented on pkg/soc/soc.go:
```diff
@@ -0,0 +1,243 @@
+
+// WithOwnerAddress provides the possibility of setting the ethereum
+// address for the owner of an update in the absence of a signer.
+func (s *Update) WithOwnerAddress(ownerAddress *Owner) *Update {
```
The shorter way of setting multiple stuff could be an `Option` struct with span and owner address fields, but this is good for me as well. Just a suggestion ;)

nolash replied:
The owner address is set in on of two ways:

* this call
* set the signer, which implicitly sets the address

I felt that having the chainable approach makes it seem more like an optional setting.

pradovic commented on pkg/soc/soc.go:
```diff
@@ -0,0 +1,243 @@
+func (s *Update) Address() (swarm.Address, error) {
+	var err error
+	hasher := swarm.NewHasher()
+	_, err = hasher.Write(s.id)
+	if err != nil {
+		return swarm.ZeroAddress, err
+	}
```
```suggestion
      if err := hasher.Write(s.id); err != nil {
		return swarm.ZeroAddress, err
	}
```

nolash replied:
terribly sorry don't like this form

pradovic commented on pkg/soc/soc.go:
```diff
@@ -0,0 +1,243 @@
+	if err != nil {
+		return swarm.ZeroAddress, err
+	}
+	_, err = hasher.Write(s.owner.address)
+	if err != nil {
+		return swarm.ZeroAddress, err
+	}
```
```suggestion

	if err := hasher.Write(s.owner.address); err != nil {
		return swarm.ZeroAddress, err
	}
```

pradovic commented on pkg/soc/soc.go:
```diff
@@ -0,0 +1,243 @@
+	if err != nil {
+		return swarm.ZeroAddress, err
+	}
+	addressBytes := hasher.Sum(nil)
+	return swarm.NewAddress(addressBytes), nil
```
```suggestion
	return swarm.NewAddress(hasher.Sum(nil)), nil
```

nolash replied:
terribly sorry but I find it give better clarity to name each step of mutations as much as possible in any code.

pradovic commented on pkg/soc/soc.go:
```diff
@@ -0,0 +1,243 @@
+	cursor += IdSize
+
+	update.signature = chunkData[cursor : cursor+SignatureSize]
+	cursor += SignatureSize
```
Could you explain to me the cursors values in this function please? 🙏

nolash replied:
the cursor points to the offset position in the byte buffer from which we read the data to be deserialized for a particular type. For example, here signature follows the chunk "id" in the serialized byte vector, and you read a slice from there until the known length of a signature.

zelig commented on pkg/soc/soc.go:
```diff
@@ -0,0 +1,243 @@
+	}, nil
+}
+
+// Update wraps a single soc update.
```
how the notion of update made its way here.
I dont think there. is update on the level of soc

zelig commented on pkg/soc/soc.go:
```diff
@@ -0,0 +1,243 @@
+	payloadSum := bmtHasher.Sum(nil)
+
+	// recover owner information
+	recoveredPublicKey, err := crypto.Recover(update.signature, payloadSum)
```
it is not the payloadSum that is signed. but the payloadSum against the id

nolash replied:
in that case I would encourage revising the figure in the book.

zelig commented on pkg/soc/soc.go:
```diff
@@ -0,0 +1,243 @@
+	payloadSum := bmtHasher.Sum(nil)
+
+	// sign the update
+	signature, err := s.signer.Sign(payloadSum)
```
it is not the payloadSum that is signed. but the payloadSum against the id

nolash replied:
... as it looks like it could also be the address it's signed with. Also wonder why there is a dotted line.

nolash commented:
Agreed, a bit of DRYing wouldn't hurt.

nolash commented:
> also not sure why separate validator pkg is needed

This is how we chose to implement validation for content addressed chunks at the time. It seems to benefit clarity to have soc validation in the same place.

jmozah commented on pkg/soc/soc.go:
```diff
@@ -0,0 +1,264 @@
+	}
+	payloadSum := bmtHasher.Sum(nil)
+	sha3Hasher := swarm.NewHasher()
+	_, err = sha3Hasher.Write(payloadSum)
```
per BOS, first write should be `ID` and then `payloadSum`

jmozah commented on pkg/soc/soc.go:
```diff
@@ -0,0 +1,264 @@
+	sch.span = int64(span)
+	cursor += swarm.SpanSize
+
+	sch.payload = chunkData[cursor:]
```
if len(chunkData) < (minChunkSize + sch.span) {
    error
}

nolash replied:
I'm not sure I understand. Minimum length is checked in L128.

jmozah replied:
after we get the spanSize, is it not good to check if we have data for that span?

zelig commented on pkg/soc/soc.go:
```diff
@@ -0,0 +1,264 @@
+	toSignBytes := sha3Hasher.Sum(nil)
+
+	// sign the update
+	signature, err := s.signer.Sign(toSignBytes)
```
i. dont think you need to hash them together, you can just use 

     toSignBytes. = bmtHash | ID

nolash replied:
Reducing to a hash before signing is a fairly common concept, though?

nolash replied:
I think it's a fairly common construct to sign digests rather than arbitrary length bytes.

In any case, the signer implementation seems to be acting up when trying to sign this concatenation. 

May I suggest that you adjust the spec to specify signature is digest instead?

@zelig

zelig replied:
yes but the `signer.Sign` does that anyway no?

nolash replied:
doesnt seem to.

nolash commented:
@zelig "rewrite the code"

ok. got you. thanks.

nolash commented:
@zelig I've added convenience accessors re your comments.

I find it pertinent to have a representation of soc which is independent of whether you have the signing key or not, for anyone wanting to manipulate this directly. I've DRY'ed the code, and also added the previously non-existent "content addressed" package which you solicited in your review by splitting it out from the package previously called validator.

zelig commented on pkg/soc/soc_test.go:
```diff
@@ -0,0 +1,130 @@
+	}
+
+	// verify owner match
+	//recoveredPublicKey, err := crypto.Recover(signature, toSignBytes)
```
commented code

pradovic commented on pkg/content/content.go:
```diff
@@ -0,0 +1,82 @@
+
+// NewContentChunk creates a new content-addressed single-span chunk.
+// The length of the chunk data is set as the span.
+func NewContentChunk(data []byte) (swarm.Chunk, error) {
```
Since the package is content, you could also just use a word Chunk instead of ContentChunk, because the usage of exported stuff will be `content.Chunk` or `content.ContentChunk`, choose whatever looks better for you :)

pradovic commented on pkg/content/validator.go:
```diff
@@ -0,0 +1,33 @@
+
+// ContentAddressValidator validates that the address of a given chunk
+// is the content address of its contents.
+type ContentAddressValidator struct {
```
Same as the above, you can ommit `Content` in naming inside `content` package if you like.

pradovic commented on pkg/soc/soc.go:
```diff
@@ -0,0 +1,234 @@
+}
+
+// CreateAddress creates a new soc address from the soc id and the ethereum address of the signer
+func CreateAddress(id Id, owner *Owner) (swarm.Address, error) {
```
`NewAddress` might be more idiomatic.

nolash replied:
I am not sure. `New` suggests - to me at least - instantiate an object, reserve memory etc. This is more like a helper function.

pradovic replied:
Makes sense!
